File: src/main/java/com/refinedmods/refinedstorage/item/UpgradeItem.java
package com.refinedmods.refinedstorage.item;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.render.Styles;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.level.Level;

import javax.annotation.Nullable;
import java.util.List;

public class UpgradeItem extends Item {
    private final Type type;

    public UpgradeItem(Type type) {
        super(new Item.Properties().tab(RS.CREATIVE_MODE_TAB));

        this.type = type;
    }

    @Override
    public void appendHoverText(ItemStack stack, @Nullable Level level, List<Component> tooltip, TooltipFlag flag) {
        super.appendHoverText(stack, level, tooltip, flag);

        if (type.getFortuneLevel() > 0) {
            tooltip.add(
                Component.translatable("enchantment.minecraft.fortune")
                    .append(" ")
                    .append(Component.translatable("enchantment.level." + type.getFortuneLevel()))
                    .setStyle(Styles.GRAY)
            );
        }
    }

    @Override
    public boolean isFoil(ItemStack stack) {
        return type == Type.SILK_TOUCH ||
            type == Type.FORTUNE_1 ||
            type == Type.FORTUNE_2 ||
            type == Type.FORTUNE_3;
    }

    public Type getType() {
        return type;
    }

    public enum Type {
        NORMAL("normal"),
        SPEED("speed"),
        RANGE("range"),
        CRAFTING("crafting"),
        STACK("stack"),
        SILK_TOUCH("silk_touch"),
        FORTUNE_1("fortune_1"),
        FORTUNE_2("fortune_2"),
        FORTUNE_3("fortune_3"),
        REGULATOR("regulator");

        private final String name;

        Type(String name) {
            this.name = name;
        }

        public String getName() {
            return name;
        }

        public int getEnergyUsage() {
            switch (this) {
                case NORMAL:
                    return 0;
                case RANGE:
                    return RS.SERVER_CONFIG.getUpgrades().getRangeUpgradeUsage();
                case SPEED:
                    return RS.SERVER_CONFIG.getUpgrades().getSpeedUpgradeUsage();
                case CRAFTING:
                    return RS.SERVER_CONFIG.getUpgrades().getCraftingUpgradeUsage();
                case STACK:
                    return RS.SERVER_CONFIG.getUpgrades().getStackUpgradeUsage();
                case SILK_TOUCH:
                    return RS.SERVER_CONFIG.getUpgrades().getSilkTouchUpgradeUsage();
                case FORTUNE_1:
                    return RS.SERVER_CONFIG.getUpgrades().getFortune1UpgradeUsage();
                case FORTUNE_2:
                    return RS.SERVER_CONFIG.getUpgrades().getFortune2UpgradeUsage();
                case FORTUNE_3:
                    return RS.SERVER_CONFIG.getUpgrades().getFortune3UpgradeUsage();
                case REGULATOR:
                    return RS.SERVER_CONFIG.getUpgrades().getRegulatorUpgradeUsage();
                default:
                    throw new IllegalStateException("What even am I?");
            }
        }

        public int getFortuneLevel() {
            switch (this) {
                case FORTUNE_1:
                    return 1;
                case FORTUNE_2:
                    return 2;
                case FORTUNE_3:
                    return 3;
                default:
                    return 0;
            }
        }
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/item/UpgradeItem.kt
package com.refinedmods.refinedstorage.item

import com.refinedmods.refinedstorage.RS
import com.refinedmods.refinedstorage.render.Styles
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.util.text.Text
import net.minecraft.util.text.TranslationTextComponent
import net.minecraft.world.World

class UpgradeItem(val type: Type) : Item(Properties().group(RS.MAIN_GROUP)) {
    enum class Type(override val name: String) {
        NORMAL("normal"), SPEED("speed"), RANGE("range"), CRAFTING("crafting"), STACK("stack"), SILK_TOUCH("silk_touch"), FORTUNE_1("fortune_1"), FORTUNE_2("fortune_2"), FORTUNE_3("fortune_3"), REGULATOR("regulator");

        val energyUsage: Int
            get() = when (this) {
                NORMAL -> 0
                RANGE -> RS.SERVER_CONFIG.upgrades.getRangeUpgradeUsage()
                SPEED -> RS.SERVER_CONFIG.upgrades.getSpeedUpgradeUsage()
                CRAFTING -> RS.SERVER_CONFIG.upgrades.getCraftingUpgradeUsage()
                STACK -> RS.SERVER_CONFIG.upgrades.getStackUpgradeUsage()
                SILK_TOUCH -> RS.SERVER_CONFIG.upgrades.getSilkTouchUpgradeUsage()
                FORTUNE_1 -> RS.SERVER_CONFIG.upgrades.getFortune1UpgradeUsage()
                FORTUNE_2 -> RS.SERVER_CONFIG.upgrades.getFortune2UpgradeUsage()
                FORTUNE_3 -> RS.SERVER_CONFIG.upgrades.getFortune3UpgradeUsage()
                REGULATOR -> RS.SERVER_CONFIG.upgrades.getRegulatorUpgradeUsage()
                else -> throw IllegalStateException("What even am I?")
            }
        val fortuneLevel: Int
            get() = when (this) {
                FORTUNE_1 -> 1
                FORTUNE_2 -> 2
                FORTUNE_3 -> 3
                else -> 0
            }
    }

    fun addInformation(stack: ItemStack?, @Nullable world: World?, tooltip: MutableList<Text?>, flag: ITooltipFlag?) {
        super.addInformation(stack, world, tooltip, flag)
        if (type.fortuneLevel > 0) {
            tooltip.add(
                    TranslationTextComponent("enchantment.minecraft.fortune")
                            .appendString(" ")
                            .append(TranslationTextComponent("enchantment.level." + type.fortuneLevel))
                            .setStyle(Styles.GRAY)
            )
        }
    }

    fun hasEffect(stack: ItemStack?): Boolean {
        return type == Type.SILK_TOUCH || type == Type.FORTUNE_1 || type == Type.FORTUNE_2 || type == Type.FORTUNE_3
    }

    init {
        this.setRegistryName(RS.ID, if (type == Type.NORMAL) "upgrade" else type.name + "_upgrade")
    }
}